src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h

```diff
@@ -0,0 +1,148 @@
+    if (cell->prev) {
+        MPIDI_POSIX_EAGER_IQUEUE_cell_t *prev =
+            MPIDI_POSIX_EAGER_IQUEUE_GET_CELL(transport, cell->prev);
+        MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(transport, prev);
+        cell->prev = 0;
+    }
+    cell->next = NULL;
+    if (transport->last_cell) {
+        transport->last_cell->next = cell;
+    } else {
+        transport->first_cell = cell;
+    }
+    transport->last_cell = cell;
+}
+
+static inline MPIDI_POSIX_EAGER_IQUEUE_cell_t
+    * MPIDI_POSIX_EAGER_IQUEUE_receive_cell(MPIDI_POSIX_EAGER_IQUEUE_transport_t * transport)
+{
+    MPIDI_POSIX_EAGER_IQUEUE_cell_t *cell = NULL;
+    MPIDI_POSIX_EAGER_IQUEUE_terminal_t *terminal;
+
+    /* Update the first_cell pointer in the queue if we pull out the first cell */
+    if (transport->first_cell) {
+        cell = transport->first_cell;
+        if (cell->next == NULL) {
+            transport->first_cell = NULL;
+            transport->last_cell = NULL;
+        } else {
+            transport->first_cell = cell->next;
+        }
+    } else {
+        /* If first_cell wasn't set, grab the next cell from the appropriate terminal */
+        terminal = &transport->terminals[transport->local_rank];
+
+        if (terminal->head) {
+            uintptr_t head =
+                (uintptr_t) (unsigned int *) OPA_swap_ptr((OPA_ptr_t *) & terminal->head, NULL);
+
+            cell = MPIDI_POSIX_EAGER_IQUEUE_GET_CELL(transport, head);
+
+            /* If the head of the terminal has the prev pointer set, enqueue the current cell and
+             * use that one instead. */
+            if (cell->prev) {
+                MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(transport, cell);
+
+                cell = transport->first_cell;
+                MPIR_Assert(cell != NULL);
+                MPIR_Assert(cell->next != NULL);
+                transport->first_cell = cell->next;
+            }
+        }
+    }
+
+    return cell;
+}
+
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_POSIX_eager_recv_begin
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int
+MPIDI_POSIX_eager_recv_begin(MPIDI_POSIX_eager_recv_transaction_t * transaction)
+{
+    MPIDI_POSIX_EAGER_IQUEUE_transport_t *transport;
+    MPIDI_POSIX_EAGER_IQUEUE_cell_t *cell;
+
+    /* Get the transport with the global variables */
+    transport = MPIDI_POSIX_EAGER_IQUEUE_get_transport();
+
+    cell = MPIDI_POSIX_EAGER_IQUEUE_receive_cell(transport);
+
+    if (cell) {
+        transaction->src_grank = transport->local_procs[cell->from];
+        transaction->payload = MPIDI_POSIX_EAGER_IQUEUE_CELL_PAYLOAD(cell);
+        transaction->payload_sz = cell->payload_size;
+
+        if (likely(cell->type == MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HEAD)) {
+            transaction->msg_hdr = &cell->am_header;
+        } else {
+            MPIR_Assert(cell->type == MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL);
+            transaction->msg_hdr = NULL;
+        }
```

Comment:
The naming of `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HEAD` and `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL` seems confusing. It actually means whether it contains an AM header, but the type names read like location in a queue...Maybe we could change them to `...CELL_TYPE_HDR` and `...CELL_TYPE_DATA` ?

Reply:
`transaction` is not initialized.

Reply:
> The naming of `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HEAD` and `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL` seems confusing. It actually means whether it contains an AM header, but the type names read like location in a queue...Maybe we could change them to `...CELL_TYPE_HDR` and `...CELL_TYPE_DATA` ?

That sounds good to me. 👍

Reply:
> `transaction` is not initialized.

Not sure what you mean here. The variable is created in [posix_progress.h](https://github.com/pmodels/mpich/pull/3199/files/8ff97b4aa32837c97d95c212e66dac64b45ae856..49167afc81da35f29869d1db4d8d0ba35cb41be3#diff-b7a55510bbf1a862e99ad3afd3df53e2R27) and passed in as a parameter. If a cell isn't found, `transaction` isn't used.

Reply:
>> transaction is not initialized.

>Not sure what you mean here. The variable is created in posix_progress.h and passed in as >a parameter. If a cell isn't found, transaction isn't used.

Yes, now I got it. Please just ignore this comment.